java/com/google/re2j/Machine.java

```diff
@@ -28,68 +26,49 @@
 
   // A queue is a 'sparse array' holding pending threads of execution.  See:
   // research.swtch.com/2008/03/using-uninitialized-memory-for-fun-and.html
-  
-  // order matters
   private static class Queue {
 
-    static class Entry {
-      int pc;
-      Thread thread;
-    }
-
-    final Entry[] dense; // may contain stale Entries in slots >= size
+    final Thread[] dense; // may contain stale Entries in slots >= size
     final int[] sparse;  // may contain stale but in-bounds values.
     int size;  // of prefix of |dense| that is logically populated
 
     Queue(int n) {
       this.sparse = new int[n];
-      this.dense = new Entry[n];
+      Arrays.fill(sparse, Integer.MAX_VALUE);
+      this.dense = new Thread[n];
     }
 
     boolean contains(int pc) {
-      int j = sparse[pc];  // (non-negative)
-      if (j >= size) {
-        return false;
-      }
-      Entry d = dense[j];
-      return d != null && d.pc == pc;
+      return sparse[pc] < size;
     }
 
     boolean isEmpty() { return size == 0; }
 
-    Entry add(int pc) {
+    int add(int pc) {
       int j = size++;
       sparse[pc] = j;
-      Entry e = dense[j];
-      if (e == null) {  // recycle previous Entry if any
-        e = dense[j] = new Entry();
-      }
-      e.thread = null;
-      e.pc = pc;
-      return e;
+      dense[j] = null;
+      return j;
     }
 
-    // Frees all threads on the thread queue, returning them to the free pool.
-    void clear(List<Thread> freePool) {
-      for(int i = 0; i < size; ++i) {
-        Entry entry = dense[i];
-        if (entry != null && entry.thread != null) {
-          // free(entry.thread)
-          freePool.add(entry.thread);
-        }
-        // (don't release dense[i] to GC; recycle it.)
-      }
+
+    void clear() {
+      Arrays.fill(sparse, Integer.MAX_VALUE);
```

Comment:
If the cost of clear is significant to running time, you might also want to benchmark using not one array of Entry =
 (thread, pc), but two arrays, one of Thread and one of pcs, and thus avoid the need to allocate Entries.  (In the original Go, the array of entries is flat, and thus avoids allocation.  I missed that when I did the port.)

Doing it that way would avoid the need to fill the entire sparse array with MAX_VALUE, which might get expensive for complex state machines.

Reply:
so in the benchmarks the perf are very similar with will or densePcs so I reverted to use the pc, as it's more consistent with the other impl